lib/connection/pool.js

```diff
@@ -215,21 +215,30 @@ function stateTransition(self, newState) {
   }
 }
 
-function authenticate(pool, auth, connection, cb) {
-  if (auth[0] === undefined) return cb(null);
+function authenticate(pool, credentials, connection, cb) {
+  // If there are no credentials, do not authenticated
+  if (!credentials) {
+    return cb(null);
+  }
+
+  // HACK: If default mechanism has not been determined yet, do not
+  // try to authenticate
+  if (credentials.mechanism === 'default') {
```

Comment:
is this doing what the comment indicates it's doing? This looks like `If the mechanism IS default, do not try to authenticate`

Reply:
The explanation for this is a little complicated:

+ If you do not provide a mechanism, or if you say "default" we need do to mechanism negotiation. As of MongoDB 4.0, this requires an `ismaster` to get `saslSupportedMechs`
+ For mechanism negotiation to work, we need to have credentials available at the time of topology connect, which happens before the actual `auth` call. This way, we can add the `saslSupportedMechs` field to the first outgoing `ismaster`
+ At the same time, there is legacy support for authenticating on connect, where you pass credentials in to the initial connect call, and immediately auth. This is a problem, since the credentials have not been mechanism negotiated yet, so they will still say `default`.

TBH I'm not happy about the workflow here, and I'd like to brainstorm some solutions for it. But this is currently mimicking the original behavior.

Reply:
ok, two things: can you update the note to better reflect what's going on here (just so I don't make the same mistake when reading through again), and secondly can you create a ticket to track refactoring this process, and add it to the Auth Refactor epic

Reply:
Issue is [here](https://jira.mongodb.org/browse/NODE-1674)